witness/golang/omniwitness/usbarmory/internal/storage/persistance.go

```diff
@@ -0,0 +1,293 @@
+	}
+	p.mapSlot = s
+	if err := p.populateMap(); err != nil {
+		return fmt.Errorf("failed to populate logID → slot map: %v", err)
+	}
+	return nil
+}
+
+// Logs returns the IDs of all logs that have checkpoints that can
+// be read.
+func (p *SlotPersistence) Logs() ([]string, error) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	r := make([]string, 0, len(p.idToSlot))
+	for k := range p.idToSlot {
+		r = append(r, k)
+	}
+	return r, nil
+}
+
+// ReadOps returns read-only operations for the given log ID. This
+// method only makes sense for IDs returned by Logs().
+func (p *SlotPersistence) ReadOps(logID string) (persistence.LogStateReadOps, error) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	i, ok := p.idToSlot[logID]
+	if !ok {
+		// TODO(al): work around undocumented assumptions in the storage interface;
+		// we have to return a ReadOp even if we don't know about the specified logID.
+		// the ReadOps we return must then return an error of codes.NotFound when GetLatest
+		// is called so that the witness will eventually call WriteOps/Update with the
+		// same logID and create the record.
+		return &slotOps{}, nil
+	}
+	s, err := p.part.Open(i)
+	if err != nil {
+		return nil, fmt.Errorf("internal error opening slot %d associated with log ID %q: %v", i, logID, err)
+	}
+	return &slotOps{slot: s}, nil
+}
+
+// WriteOps shows intent to write data for the given logID. The
+// returned operations must have Close() called when the intent
+// is complete.
+// There is no requirement that the ID is present in Logs(); if
+// the ID is not there and this operation succeeds in committing
+// a checkpoint, then Logs() will return the new ID afterwards.
+func (p *SlotPersistence) WriteOps(logID string) (persistence.LogStateWriteOps, error) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	i, ok := p.idToSlot[logID]
+	if !ok {
+		var err error
+		i, err = p.addLog(logID)
+		if err != nil {
+			glog.V(2).Infof("Failed to add mapping: %v", err)
+			return nil, fmt.Errorf("unable to assign slot for log ID %q: %v", logID, err)
+		}
+		glog.V(2).Infof("Added new mapping %s -> %d", logID, i)
+	}
+	glog.V(2).Infof("mapping %s -> %d", logID, i)
+	s, err := p.part.Open(i)
+	if err != nil {
+		glog.V(2).Infof("failed to open %d: %v", i, err)
+		return nil, fmt.Errorf("internal error opening slot %d associated with log ID %q: %v", i, logID, err)
+	}
+	return &slotOps{slot: s}, nil
+}
+
+type slotOps struct {
+	mu         sync.Mutex
+	slot       *slots.Slot
+	writeToken uint32
+}
+
+type logRecord struct {
+	Checkpoint []byte
+	Proof      []byte
+}
+
+// GetLatest returns the latest checkpoint and its compact range (if applicable).
+// If no checkpoint exists, it must return codes.NotFound.
+func (s *slotOps) GetLatest() ([]byte, []byte, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// TODO(al): workaround for storage assumption - see comment in ReadOps above.
+	if s.slot == nil {
+		return nil, nil, status.Error(codes.NotFound, "no checkpoint for log")
+	}
+
```

Comment:
As noted above, I think this should be removed and just have a different type that is for an unknown log. Then `s.slot` should never be nil.

Reply:
as above, will leave this as is for now and we'll figure out how it should all hang together and update.